src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersState.scala

```diff
@@ -5,83 +5,165 @@ import com.typesafe.scalalogging.StrictLogging
+    hungryInstances.getOrElse(role, Map.empty).contains(instanceId)
+  }
+
+  private def updateInstanceState(role: String, instanceId: Instance.Id, newState: Option[Instance]): ReviveOffersState = {
+    val newHungryState = newState.filter(isHungry).map(toHungryInstance)
+
+    val newHungryInstances: Map[String, Map[Instance.Id, HungryInstance]] = newHungryState match {
+      case Some(hungryInstance) =>
+        hungryInstance.reason match {
+          case ReviveReason.Launching =>
+            logger.debug(s"Adding ${instanceId} to scheduled instances.")
+          case ReviveReason.CleaningUpReservations =>
+            logger.debug(s"$instanceId is terminal but has a reservation.")
+        }
+        val newRoleHungryInstances = hungryInstances.getOrElse(role, Map.empty) + (instanceId -> hungryInstance)
+        hungryInstances + (role -> newRoleHungryInstances)
+      case None =>
+        if (hasInstance(role, instanceId))
+          logger.debug(s"Removing ${instanceId} from instances wanting offers.")
+        val newRoleHungryInstances = hungryInstances.getOrElse(role, Map.empty) - instanceId
+        val newHungryInstances =
+          if (newRoleHungryInstances.isEmpty)
+            hungryInstances - role
+          else
+            hungryInstances + (role -> newRoleHungryInstances)
+
+        newHungryInstances
+    }
+
+    val newAllInstanceIdsByRole = if (newState.nonEmpty) {
+      val withInstanceIdAdded = allInstanceIdsByRole.getOrElse(role, Set.empty) + instanceId
+      allInstanceIdsByRole + (role -> withInstanceIdAdded)
+    } else {
+      val withInstanceIdRemoved = allInstanceIdsByRole.getOrElse(role, Set.empty) - instanceId
+      if (withInstanceIdRemoved.isEmpty)
+        allInstanceIdsByRole - role
+      else
+        allInstanceIdsByRole + (role -> withInstanceIdRemoved)
+    }
+
+    update(hungryInstances = newHungryInstances, allInstanceIdsByRole = newAllInstanceIdsByRole)
   }
 
   /** @return this state updated with an instance. */
   def withInstanceUpdated(instance: Instance): ReviveOffersState = {
-    logger.debug(s"${instance.instanceId} updated to ${instance.state}")
-    if (instance.isScheduled) {
-      logger.debug(s"Adding ${instance.instanceId} to scheduled instances.")
-      copy(scheduledInstances = scheduledInstances.updated(instance.instanceId, instance.runSpec.configRef))
-    } else if (shouldUnreserve(instance)) {
-      logger.debug(s"$instance is terminal but has a reservation.")
-      copy(scheduledInstances = scheduledInstances - instance.instanceId, terminalReservations = terminalReservations + instance.instanceId)
-    } else if (!instance.isScheduled) {
-      logger.debug(s"Removing ${instance.instanceId} from scheduled instances.")
-      copy(scheduledInstances = scheduledInstances - instance.instanceId)
-    } else this
+    updateInstanceState(instance.role, instance.instanceId, Some(instance))
+  }
+
+  private def isHungry(instance: Instance): Boolean = {
```

Comment:
As much as a like it this should probably be `requiresRevive` cause I'm not sure `isHungry` fits `shouldUnreserve`.

Reply:
Went with wantsOffers instead